app/questionnaire/questionnaire_store_updater.py

```diff
@@ -159,27 +161,30 @@ def remove_list_item_data(self, list_name: str, list_item_id: str) -> None:
             self._update_relationship_question_completeness(list_name)
 
         self._progress_store.remove_progress_for_list_item_id(list_item_id=list_item_id)
-        self._capture_dependencies_for_list_item_removal(list_name)
+        self._capture_dependencies_for_list_change(list_name)
 
-    def _capture_dependencies_for_list_item_removal(self, list_name: str) -> None:
+    def _capture_dependencies_for_list_change(self, list_name: str) -> None:
         """
-        If an item is removed from a list, dependencies of any child blocks need to be captured and their progress updated.
-        (E.g. dynamic-answers depending on a child block could have validation, so need revisiting if an item is removed)
+        Captures the dependencies when an item is added to or removed from the given list.
+        Any list collector sections will be affected by the change as well as other sections using when rules
         """
-        for list_collector in self._schema.get_list_collectors_for_list(
-            for_list=list_name
-        ):
-            child_blocks = (
-                list_collector.get("add_block"),
-                list_collector.get("remove_block"),
-                *list_collector.get("repeating_blocks", []),
-            )
-            for child_block in child_blocks:
-                if child_block:
-                    for answer_id in self._schema.get_answer_ids_for_block(
-                        child_block["id"]
-                    ):
-                        self._capture_block_dependencies_for_answer(answer_id)
+        self._capture_block_dependencies_for_list(list_name)
+        section_ids = self._schema.get_when_rule_section_dependencies_for_list(
+            list_name
+        )
+        section_ids.update(
+            self._schema.list_collector_section_ids_by_list_name[list_name]
+        )
+
+        for section_key in self.started_section_keys(section_ids=section_ids):
+            # Only add sections which are repeated sections for this list, or the section in which this list is collected
+            # Prevents list item progresses being added as dependants as these are captured by started_section_keys(section_ids=section_ids)
+            if (
+                not self._schema.get_repeat_for_section(section_key.section_id)
+                and section_key.list_item_id
+            ):
+                continue
+            self.dependent_sections.add(DependentSection(**section_key.to_dict()))
```

Comment:
Not necessarily related to this change, but could use SectionKey in the DependentSection class, rather than passing the individual vars to keep things consistent. May be overkill though

Reply:
As in 
```python
class DependentSection(NamedTuple):
    section_key: SectionKey
    is_complete: bool | None = None
```
?
I agree that'd be nice here but I worry it'd make other initialisations of `DependentSection` worse, most of the use cases aren't with a section_key. I'd probably lean the other way in this case

Reply:
Yeah that's what I meant, yeah I can see lots of places where it might be slightly worse, but kind of made sense to me as dependent section has a method to return a section key as well. Only other thing is that if we made this change there would be barely any difference between the DependentSection and SectionKey classes.

Reply:
Open for some team views on this - I have tried adding this throughout locally, I'm not keen to keep it due to the number of instances where we'd have something like `DependentSection(SectionKey(section_id, list_item_id))` 
Where repeating section dependencies are done by iterating through list store and using list item with section, so the two aren't already paired by a section key. 
But there isn't that much in it and if it makes more sense to everyone and they prefer it for following dependencies I can go with it

Reply:
I think in this specific bit, passing section key does look like a nice win-win but I agree that most initialisation don't have a section key and would therefore be worse of for readability as they become

```py
DependentSection(
	SectionKey(section_id, list_item_id)
)
```

So in this instance, my preference is as is.